### PR TITLE
[bug] All_gather support tensor on cpu

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #<issue_number>

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/apply_func.py [hunk 1]
```diff
@@ -166,4 +166,11 @@ def convert_to_tensors(data, device: torch.device = None):
         raise MisconfigurationException("device (torch.device) should be provided.")
     for src_dtype, conversion_func in CONVERSION_DTYPES:
         data = apply_to_collection(data, src_dtype, partial(conversion_func, device=device))
+
+    def _move_to_device(value, device=None):
+        if device is not None and value.device != device:
+            value = value.to(device)
+        return value.contiguous()
+
```
Comment (ananthsub) [inline pytorch_lightning/utilities/apply_func.py:174]: - please add typehints
- in L165, if we raise if `device` is `None`, why do we set `None` as the default vs making it a required argument?

Comment (tchaton) [reply pytorch_lightning/utilities/apply_func.py:174]: You are right ! Let me update this.

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/utilities/apply_func.py [hunk 2]
```diff
@@ -164,6 +164,14 @@ def batch_to(data):
...
     if device is None:
         raise MisconfigurationException("device (torch.device) should be provided.")
+
     for src_dtype, conversion_func in CONVERSION_DTYPES:
         data = apply_to_collection(data, src_dtype, partial(conversion_func, device=device))
+
+    def _move_to_device_and_make_contiguous(t: torch.Tensor, device: torch.device):
+        if t.device != device:
+            t = t.to(device)
+        return t.contiguous()
```
Comment (carmocca) [inline pytorch_lightning/utilities/apply_func.py:174]: No need for the if

```suggestion
        return t.to(device).contiguous()
```

File: tests/utilities/test_all_gather_grad.py [hunk 3]
```diff
@@ -58,13 +58,17 @@ def training_epoch_end(self, outputs) -> None:
             self.training_epoch_end_called = True
             losses = torch.stack([x["loss"] for x in outputs])
             gathered_loss = self.all_gather({
+                "losses_tensor_int": torch.tensor([1, 2, 3]),
+                "losses_tensor_float": torch.tensor([1., 2., 3.]),
```
Comment (awaelchli) [inline tests/utilities/test_all_gather_grad.py:62]: but these tensors are contiguous already. You need a multi dimensional tensor, and you can easily make it non-contiguous by adding a transpose, like this:

`torch.rand(2,2).t()`

Comment (tchaton) [reply tests/utilities/test_all_gather_grad.py:62]: Thanks !

File: tests/utilities/test_all_gather_grad.py [hunk 4]
```diff
@@ -58,13 +58,17 @@ def training_epoch_end(self, outputs) -> None:
...
+                "losses_tensor_float": torch.rand(2, 2).t(),
                 "losses_np_ndarray": np.array([1, 2, 3]),
                 "losses_bool": [True, False],
                 "losses_float": [0., 1., 2.],
                 "losses_int": [0, 1, 2],
                 "losses": losses,
                 "losses_list": [losses, losses]
             })
+            assert gathered_loss["losses_tensor_int"][0].dtype == torch.int64
+            assert gathered_loss["losses_tensor_float"][0].dtype == torch.float
```
Comment (awaelchli) [inline tests/utilities/test_all_gather_grad.py:71]: here also check the device?

**Review by awaelchli — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]
